docassemble/ALToolbox/al_income.py

```diff
@@ -127,21 +142,20 @@ def __str__(self) -> str:
     def total(self, times_per_year: float = 1) -> Decimal:
         """
         Returns the income over the specified times_per_year, taking into account
-        hours per period for hourly items.
+        hours per period for hourly items. For example, for an hourly income of 10
+        an hour, 40 hours a week, `income.total(1)` would be 20,800, the yearly income,
+        and `income.total(52)` would be 400, the weekly income.
 
         To calculate `.total()`, an ALIncome must have a `.times_per_year` and `.value`.
         It can also have `.is_hourly` and `.hours_per_period`.
         """
+        val = _currency_float_to_decimal(self.value)
         if hasattr(self, "is_hourly") and self.is_hourly:
             return (
-                Decimal(self.value)
-                * Decimal(self.hours_per_period)
-                * Decimal(self.times_per_year)
+                val * Decimal(self.hours_per_period) * Decimal(self.times_per_year)
             ) / Decimal(times_per_year)
         else:
-            return (Decimal(self.value) * Decimal(self.times_per_year)) / Decimal(
-                times_per_year
-            )
+            return (val * Decimal(self.times_per_year)) / Decimal(times_per_year)
```

Comment:
Math isn't my strong suit, so I'm thinking out loud here. Could this still return a value that doesn't conform to the locale format? That said, would we even want this to always return a value that conforms to the locale format or do we want developers to use `currency` for that? Though `val` should be (and is given these changes) a locale-formatted value to get the user input correct, maybe the rest of the math here should return a more accurate fraction so that further math operations could be based on the most accurate values.

Reply:
Good catch! I had been assuming that users would use `currency()` for that. There are some ways that we could make it always return a value with exactly `fraction_digits` decimal places, but I'm not sure we'd want to do that? I'm unsure how exactly this sort of accounting math is done, but it would only really happen when you have non-standard times, i.e. if you get paid monthly, what would you say your weekly pay is? (that question doesn't make too much sense IMO)  Or, if you get paid bi-weekly, what is your monthly pay? Does it depends on the month (we're not being that detailed with this library to answer it that way), or do you average things out?

I'm leaning towards leavings things as accurate as possible, since the results of these functions could be used in other equations; we don't want to round too early. For now I added a test that uses `Decimal::quantize` (the same function we would use to address this) to check that at least when rounding the value is still correct.

Reply:
Pinging @nonprofittechy, I'm not sure if you could answer any of these accounting questions.

Reply:
Where this would come up would be in income qualification questions. A user won't generally be penalized for averaging it slightly differently (e.g., the 30.416 days vs a 28 day month) if the form asks for "monthly income". I've never seen a form ask for "weekly" income except to help the user report it accurately--they're likely to ask for monthly or annual income. Keep in mind that for most of our client populations income can vary quite a lot from month to month.

For the real-world context, often, the user reports the approximate value on the form and the organization might ask for the last 4 pay stubs + most recent tax return and do their own math, but they want the approximate answer on the form.

I would suggest we do the possible rounding as close to the end of the calculation as possible so our default is more precision.